src/layers/singleLayer.jl

```diff
@@ -62,62 +63,65 @@ function initTheta(this::singleLayer{T}) where {T <: Number}
 end
 
 
-function Jthetamv(this::singleLayer{T},dtheta::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
-    dA             = tmp[2]
-    nex            = div(length(Y),nFeatIn(this))
-    Y              = reshape(Y,:,nex)
+function Jthetamv(this::singleLayer{T},dtheta::Array{T},theta::Array{T},Yin::Array{T},tmp) where {T <: Number}
+    dA::Array{T,2}             = tmp[2]
+    nex            = div(length(Yin),nFeatIn(this))
+    Y              = reshape(Yin,:,nex)
 
     th1,th2,th3,th4     = splitWeights(this,theta)
     dth1,dth2,dth3,dth4 = splitWeights(this,dtheta)
-    dZ = Jthetamv(this.K,dth1,th1,Y) .+ this.Bin*dth2
+    dZ::Array{T,2} = Jthetamv(this.K,dth1,th1,Y) .+ this.Bin*dth2
     Kop = getOp(this.K,th1)
     dZ  = Jmv(this.nLayer,dth4,dZ,th4,Kop*Y.+this.Bin*th2,tmp[1])[2]
-    dZ = dA.*dZ .+ this.Bout*dth3;
+    dZ .*= dA 
+    dZ .+= this.Bout*dth3
     return dZ, dZ
 end
 
-function JYmv(this::singleLayer{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
+function JYmv(this::singleLayer{T},dYin::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
     dA  = tmp[2]
-    nex = div(length(dY),nFeatIn(this))
+    nex = div(length(dYin),nFeatIn(this))
     th1,th2,th3,th4 = splitWeights(this,theta)
     Kop         = getOp(this.K,th1)
-    dY          = reshape(dY,:,nex)
+    dY          = reshape(dYin,:,nex)
     dZ          = Kop*dY
     dZ          = JYmv(this.nLayer,dZ,th4,Kop*Y.+this.Bin*th2,tmp[1])[2]
-    dZ          = dA.*dZ
+    # dZ          = dA.*dZ
+    dZ .*= dA
     return dZ,dZ
 end
 
-function Jmv(this::singleLayer{T},dtheta::Array{T},dY::Array{T},theta::Array{T},Y::Array{T},tmp) where {T <: Number}
-    dA  = tmp[2]
-    nex = div(length(Y),nFeatIn(this))
+function Jmv(this::singleLayer{T},dtheta::Array{T},dYin::Array{T},theta::Array{T},Yin::Array{T},tmp) where {T <: Number}
+    dA::Array{T,2}  = tmp[2]
+    nex = div(length(Yin),nFeatIn(this))
     th1,th2,th3,th4 = splitWeights(this,theta)
     dth1,dth2,dth3,dth4 = splitWeights(this,dtheta)
 
-    dY = reshape(dY,:,nex);
+    dY = reshape(dYin,:,nex);
     Kop = getOp(this.K,th1)
-    dZ = Kop*dY;
+    dZ::Array{T, 2} = Kop*dY;
```

Comment:
Other fix is to call `Amv(this.K, th1, dY)` explicitly until we figure out whats going on with the LinearOperators. 

Not sure if there is any performance difference, but if it's possible for dZ to not be an Array{T, 2} it will be a more robust method